### DISCUSSION TITLE
Using optionals instead of exceptions

### DISCUSSION THREAD

**Discussion by Netwyn** [HypixelDev/PublicAPI#510] [Ideas]
Instead of throwing exceptions, optionals (values that may/may not exist stuffed inside of a Java class) could be returned.

https://docs.oracle.com/javase/8/docs/api/java/util/Optional.html
https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/CompletableFuture.html

**Comment by TheNullicorn** [comment]
Which part of the Java wrapper are you referring to? I'd be happy to start a PR for it if something's weird.

Also sorry it's not too clear, but issues with the Java implementation can go in the [Issues](https://github.com/HypixelDev/PublicAPI/issues) section. Discussions are more for general help with the API in any language.

**Reply by Netwyn** [reply]
I've noticed that most methods throw exceptions a lot, and instead of doing this, I think Optionals would be a far easier way to accomplish the same task in a more readable manner.

**Reply by Netwyn** [reply]
Also, yes, thanks for the info, I will post in the Issues section if I find something like this again.

**Reply by ConnorLinfoot** [reply]
It would depend on why the exception is being thrown, would need a more specific example really. But optionals are better seen as an alternative to a nullable method. Exceptions are the correct way if the state is not correct, such as an invalid status code.

Optionals are also a wrapper object, so unless you expect it to be missing just as much as existing its usually better to avoid them I believe, at least when performance is a high priority.

**Reply by TheNullicorn** [reply]
Looking through all the `throw`s in the library I don't see anything out of the ordinary. The Apache transport throws some RuntimeExceptions that could be UncheckedIOExceptions, but the rest seem valid to me. If you could list some of the methods you're referring to that would be helpful.

Other than that I could maybe see Optionals being used for values in Reply classes, but I feel like that would be a lot to change at once, not to mention backwards compatibility issues.
